### PR TITLE
An array as a root node

### PR DESCRIPTION
**Pull request by frenzzy**
An alternative solution for #590 and #568 POC

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -55,6 +55,11 @@ export function app(state, actions, view, container) {
 
     var next = view(globalState, wiredActions)
     if (container && !renderLock) {
+      next = {
+        nodeName: container.nodeName.toLowerCase(),
+        attributes: {},
+        children: [].concat(next)
```
Comment (frenzzy) [inline src/index.js]: or `next.pop /* Array */ ? next : [next]`

Comment (chicoxyzzy) [reply src/index.js]: Why not `Array.isArray(next)` (IE 9 compatible)?

**Comment by JorgeBucaran** [comment]
Thanks, @frenzzy. This is an interesting proposal. What about adding Fragments to this PR as well?

File: src/index.js (hunk 1 above)
Comment (frenzzy) [reply src/index.js]: because of bytes and perfomance? see 4930b5ecd06d9d5c5d6fe7311769020d313fcd6c

**Comment by frenzzy** [comment]
Fragments is a separate feature which is not quite related to the current single root node problem, I would prefer to create a separate commits (PRs) for each one. And for Features I see three different possible implementations:
1. Support VNode with empty name in patch function
2. Support for nested arrays inside `children` arrays
3. Converting fragments (nested arrays) to plain `children` arrays inside `h` function

Maybe need to create three PRs and compare..

File: src/index.js (hunk 1 above)
Comment (chicoxyzzy) [reply src/index.js]: Size over correctness... is bad thing

Comment (lukejacksonn) [reply src/index.js]: Then you might not like this repo...

Comment (JorgeBucaran) [reply src/index.js]: @chicoxyzzy That's a bit too dogmatic. We take a lighter approach. If we can get away with it, I suggest we do it. Not always, but if we can why not? :)

**Comment by frenzzy** [comment]
Lol, JSX Fragments already supported, see #596
And this PR makes the life easier to use them :)

**Comment by JorgeBucaran** [comment]
I would prefer not. Put everything in a single PR. Thanks.

**Comment by JorgeBucaran** [comment]
This is clever, but if we introduce Fragments someday in the future, I'd prefer this to be properly implemented using DocumentFragment and going all the way as to allow Components to return keyed fragments, not simply an array.

**Comment by frenzzy** [comment]
The PR is not about fragments directly, but about an ability to render more than one element into the `document.body`. I see your point.

**Comment by JorgeBucaran** [comment]
Yes, you are right. My apologies for asking you to mix two unrelated things.

**Comment by frenzzy** [comment]
+4 bytes^^

**Comment by JorgeBucaran** [comment]
The only way we'll support returning an array from the root is if we add fragment support someday, but not like this. Thanks.

**Comment by frenzzy** [comment]
The main purpose why I did this, I want that users do not try to put various elements in the container and look for a workaround for "which element Hyperapp should revive" like here #590 and #568

**Comment by JorgeBucaran** [comment]
@frenzzy Because we may add fragment support in the future, I think the best we can do now is nothing. So, both #590 and #598 may have to wait as well. Although, I like #598 and it's also what I want to do for Picodom 2.0.

- https://github.com/picodom/picodom/pull/96

**Comment by jorgebucaran** [comment]
@frenzzy Do you think we could implement this in a different way?

**Comment by infinnie** [comment]
But btw, we’ve been longing for this feature since the times of 0.1x.
Are keyed elements supported now in root-level array elements?

**Comment by jorgebucaran** [comment]
> Are keyed elements supported now in root-level array elements?

No, how would you implement that?

**Comment by frenzzy** [comment]
@jorgebucaran what exactly you dislike in the current implementation?

Since we have subviews/lazy components now, I think it make sense to change `patch` signature to accept a parent element and already resolved an array of children VNodes. And patch function will just modify all children elements without touching the parent element. What do you think?

**Comment by jorgebucaran** [comment]
> And patch function will just modify all children elements without touching the parent element. What do you think?

Do you have some code or pseudo code to explain what you mean?

**Comment by frenzzy** [comment]
Something like `patch(container, nextVDomTree, prevVDomTree)` and inside patch do not touch `container` itself, only `container.childNodes`

**Pull request closed without merging**